[RFC] Utilities: strequal() and os_getenv()

splinterofchaos opened the pull request:
This PR implements some utility functions for strings that had been briefly discussed in another PR:

| was | now |
| --- | --- |
| strcmp(a,b) == 0 | strequal(a,b) |
| stricmp(a,b) == 0 | striequal(a,b) |
| strncmp(a,b,s) == 0 | strnequal(a,b,s) |
| strnicmp(a,b,s) == 0 | strniequal(a,b,s) |
| a == NULL || *a == NUL | strempty(a) |

Maybe it could even be just `equal()`, but I feel `strequal()` is unambiguous. Using `striequal()` for case insensitive comparison is simple, but maybe `strcaseequal()` could be used instead if `striequal()` seems odd.

To avoid turning this into a monster of a PR, I only modified the code in `eval.c` to use these functions.

For `os_getenv()`, noting that almost every call checks that the variable isn't empty, I made this the default behaviour. `os_env_exists()` can now be used to check that the variable exists, even if empty. I made this change throughout the codebase.

Commit:
strings: Utilities for string equality/emptiness.

Add strequal, striequal, strnequal, strniequal to strings.s and refactor
eval.c to use them.

Also, remove vim_stricmp as variants (str(n)icmp, _str(n)icmp, or
str(n)casecmp) are available on all supported platforms.

aktau commented:
Quick scan, look pretty good. Though I never had a problem with the `strcmp(...) == 0` idiom, possibly because I'm so used to it by now.

equalsraf commented:
@splinterofchaos :+1 the `!strcmp()` and the double check for env vars annoys me profusely.

LGTM.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -45,6 +45,41 @@
+
+/// Compares two sized strings for equality.
+bool strnequal(const char *a, const char *b, size_t size)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
```
Should not all this be `FUNC_ATTR_CONST` in place of `FUNC_ATTR_PURE`?

aktau replied:
No: http://www.nongnu.org/c-prog-book/online/x1019.html

In effect, dereferencing the pointer arguments `a` and `b` (which `strcmp()` does by necessity) is equal to "reading global memory" from gcc's point of view. `const` functions are thus only for the case where no pointers are dereferenced (though it doesn't hurt to just read the pointer values themselves).

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -45,6 +45,41 @@
+bool striequal(const char *a, const char *b)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
+{
+  return strcasecmp(a, b) == 0;
```
Why `strcasecmp`? I do not see it in the C99 standard and given macros like `HAVE_STRCASECMP` it looks like NeoVim supports platforms without this function. `man strcasecmp` says that it is in “4.4BSD, POSIX.1-2001” standards (`strcmp` is in “SVr4, 4.3BSD, C89, C99”).

ZyX-I replied:
Thus I am suggesting usual `STRN?ICMP` macros.

aktau replied:
On that accord: http://botsikas.blogspot.de/2011/12/strcasecmp-identifier-not-found-when.html

```c
#if defined(_WIN32) || defined(_WIN64) 
  #define snprintf _snprintf 
  #define vsnprintf _vsnprintf 
  #define strcasecmp _stricmp 
  #define strncasecmp _strnicmp 
#endif
```

For an older version of MSVC. Don't know what the 2015 state is.

Our other (wannabe) supported POSIX-like systems should have it: *BSD and Illumos.

ZyX-I replied:
Then you should remove

    check_function_exists(strcasecmp HAVE_STRCASECMP)
    check_function_exists(strncasecmp HAVE_STRNCASECMP)

from `config/CMakeLists.txt` if you think they are defined always.

aktau replied:
I cannot find them being supported even in MSVC 2015 (possibly @equalsraf could give the final OK, but I hear his dev environment is a bit messy atm). So the function check macro's do still make sense. Possibly in such a way:

```c
#if defined(HAVE_STRCASECMP)
  #define XSTRCASECMP strcasecmp
#elif defined(HAVE_STRICMP)
  #define XSTRCASECMP _stricmp
#else
  #error "Your compiler or runtime environment does not support strcasecmp()/stricmp()."
#endif
```

Though I also think it looks really ugly.

ZyX-I replied:
It seems that on Windows NeoVim will use its own `vim_stricmp` implementation in place of `strcasecmp` (not present), `_stricmp` (never used) and `stricmp` (used if `HAVE_STRICMP`, but this is never defined).

ZyX-I replied:
@aktau There are already `STRICMP` and `STRNICMP` defines, no need to create `XSTRCASECMP`. Just make sure that `HAVE_STRICMP` define makes sense: I do not see a CMake check for it.

ZyX-I replied:
@aktau And third condition should not be a error, it should be `vim_stricmp`. Not sure whether `stricmp` (without underscore) needs to be kept: I never saw this function, but apparently Vim authors did.

aktau replied:
So:

`strcasecmp`: modern POSIX function, not supported by Windows
`stricmp`: Windows function that appears to be neither POSIX nor ANSI, despite what some blog posts say, it is [deprecated by Microsoft](https://msdn.microsoft.com/en-us/library/ms235365.aspx). Strangely enough, even Microsoft refers to these functions as being POSIX.
`_stricmp`: Windows function that's apparently the non-deprecated version of `strcmp`.

Btw, [this SO post](http://stackoverflow.com/questions/12413585/difference-of-stricmp-and-stricmp-in-visual-studio) claims that `stricmp` is merely an alias to `stricmp` in MSVC, which wouldn't surprise me.

I woudl go for `ifdef`ing towards the underscore version though, since it's explicitly referenced as non-dreprecated.

aktau replied:
> @aktau And third condition should not be a error, it should be vim_stricmp.

`strcasecmp` is supported on Linux, FreeBSD, OpenBSD, DragonFlyBSD and Illumos (checked by finding the relevant docs on Google). `_stricmp` is supported on Windows. These are all of our supported platforms and then some. I would actually advocate that we remove `vim_stricmp()` from our codebase.

If anyone bumps into the `#error`, we'll get an issue and we can either re-instate `vim_stricmp()` or find the case insensitive comparison function for that platform instead.

ZyX-I replied:
@aktau It makes sense since current implementation of `vim_stricmp` is actually not correct:

1. it is not correct to compare unicode strings just by lowercasing them;
2. they work only for 8-bit encodings.

aktau replied:
That said, I'm not sure if `strcasecmp` nor `_stricmp` are unicode-safe either. They don't perform unicode normalization as far as I am aware: http://stackoverflow.com/questions/11635/case-insensitive-string-comparison-in-c

This could at least be commented on our macro.

splinterofchaos replied:
I think it might make just as much sense to use the `STRCMP()` family of macros from `vim.h`, and maybe even move them to `strings.h` since they basically already implement this feature check. Though I am curious to hear if @equalsraf has any comments pertaining to the versions of MSVC we support.

aktau replied:
If we go this way I would definitely advocate for `strings.h`, we should slim down `vim.h` as much as possible, it is a never ending source of pain.

ZyX-I replied:
So best idea is splitting them into a) ascii_str\*cmp (used in most cases, including `==#`/`!=#`(these operators are used to make sure that strings are (not) identical) and `[<>]=?#`) and b) something based on ICU for `[<>]=?\?` and `[!=]=\?` and rare occasions (if any) in the C code, also for `strcmp()` VimL function because `#` variants all use byte comparisons (for compatibility with plugins which use strings as byte strings: there are some)?

splinterofchaos replied:
I'm looking at getting rid of `vim_str(n)icmp()`, but it actually turns out that `HAVE_STRN(I)CMP` and `_STR(N)CASECMP` aren't defined on my setup! Will have to investigate that, but in the meanwhile, could anyone check if it's the same on their system?

ZyX-I replied:
@splinterofchaos As I said there is no CMake check for `HAVE_STRNCMP`. So it will never be defined.

equalsraf replied:
@aktau the extra underscore is common practice in MSVC they mark the non underscore version as deprecated and throw a warning for the the original ones (even for [close, write, read](https://github.com/equalsraf/neovim/blob/tb-win32-any/src/nvim/os/win_defs.h#L55)) - more [here](http://stackoverflow.com/questions/23477746/what-are-the-posix-like-functions-in-msvcs-c-runtime) and a [full list](https://msdn.microsoft.com/en-us/library/ms235384.aspx).

As you said _stricmp seems to be the only way for Windows, assuming the locale is set. There is also a _stricmp_l(str, str, locale_t) and _mbsicmp.

Encoding correctness is trickier - most of the cases above compare strings from eval with string literals, the former use &encoding the later are compiler specific. Assuming everyone does the "right thing TM", I didnt expect this PR to tackle those - but maybe its time to add unittests for that.

splinterofchaos replied:
> As I said there is no CMake check for HAVE_STRICMP.

Ok, I misread that and added the checks.

ZyX-I commented on src/nvim/strings.c:
```diff
@@ -45,6 +45,41 @@
+  FUNC_ATTR_PURE
+{
+  return s == NULL || s[0] == '\0';
+}
```
And how about making this and `strn?equal` functions `static inline` in the header? This way compiler will be able to optimize them out (not suggesting for `strn?iequal` because I do not see gcc optimizing even `strcasecmp` out: most likely due to it being locale-dependent I think).

aktau replied:
I've thought about the same as well. I could be swayed either way but think for such small functions it could definitely be appropriate.

splinterofchaos replied:
:+1: -- it caused some unexpected `#include` errors, which is why I didn't do that at first, but they're fixed, now.

splinterofchaos commented:
>  I never had a problem with the strcmp(...) == 0 idiom

Ah, yes. The motivating example was actually `!strcmp(...)` as noted by @equalsraf.

Pyrohh commented on src/nvim/strings.c:
```diff
@@ -45,6 +45,41 @@
+  return strncmp(a, b, size) == 0;
+}
+
+/// Compares two strings for equality ignoring differences in case.
```
Maybe `Case insensitively compare two strings ...` (same for other `str(n)i*` functions if you agree with the suggestion). Also perhaps the param names across functions should be `s1`, `s2`, and `len`/`name`, to be more familiar to those who've read the man pages of their POSIX equivalents.

splinterofchaos replied:
`$ man 3 strcasecmp` actually says for me: "compare two strings ignoring case".

Pyrohh replied:
> $ man 3 strcasecmp actually says for me: "compare two strings ignoring case".

I didn't mean to imply otherwise; it was a suggestion I thought up independently. Anyways, up to you as they're both clear.

Pyrohh commented on src/nvim/mbyte.c:
```diff
@@ -3406,22 +3406,23 @@ static int enc_alias_search(char_u *name)
+  const char *s;
 # ifdef HAVE_NL_LANGINFO_CODESET
-  if ((s = nl_langinfo(CODESET)) == NULL || *s == NUL)
+  if (strempty(s = nl_langinfo(CODESET)))
```
Perhaps the nl_langinfo call could be moved out of the if, same with line 3417.

splinterofchaos replied:
It actually makes sense to me, considering the below branch/assignments.

Pyrohh replied:
It makes sense to me as well, but I still think it would be clearer as the assignments below don't involve function calls inside arguments in addition to the if. Up to you.

Pyrohh commented on src/nvim/os/env.c:
```diff
@@ -28,11 +28,20 @@
 #endif
 
 const char *os_getenv(const char *name)
```
This should be documented now that the behavior differs from `getenv`.
I think a reference such as `@see {os_unsetenv}` would be useful.

splinterofchaos commented:
Rebased and moved the commits around. I decided `memory.h` would be a better place for these functions since `strings.h` mostly deals with `char_u`--`memory.h` has the real meat of our string functions and pairs well with the related `mem*` ones.

Also, while I'm in the area, I thought I'd take care of #1146 since that's been bugging me for a long time. I'll turn this back to `RFC` when I've finished.

Commit:
strings: use standard functions over x-variants

Have cmake check for strdup, strlcpy, etc., and replace our x-variants
with them.

splinterofchaos commented:
Perhaps not the cleanest solution, but I added some checks to cmake for `strdup`, `strlcpy`, `stpcpy`, etc., and added `macros to `memory.c` to convert out `x` versions to the standard variants. Getting rid of the `x` part might look nicer, but since MSVC, and maybe others, already define and deprecate some of them, we might need to map `xfunc` to `_func`, and not (re-)define the original name.

Pyrohh commented on src/nvim/mbyte.c:
```diff
@@ -3406,22 +3406,23 @@ static int enc_alias_search(char_u *name)
-          s = (char *)os_getenv("LANG");
+      if ((s = os_getenv("LC_ALL")))
+        if ((s = os_getenv("LC_CTYPE")))
+          s = os_getenv("LANG");
```
Perhaps braces should be added, given git thinks this is a whole new block.

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/os/env.c:
```diff
@@ -416,11 +421,6 @@ static char *remove_tail(char *p, char *pend, char *name)
-      && *kos_env_path == NUL) {  // empty is the same as not set
-    kos_env_path = NULL;
-  }
-
```
:+1: This is really great.

Pyrohh commented:
This looks pretty good to me; could you rebase this whenever you get the chance?

splinterofchaos commented:
rebased

splinterofchaos commented:
Marking RDY. I plan to merge this tomorrow.

Pyrohh commented:
Could you rebase once more so people can review the "final product", given that there's merge conflicts.

Pyrohh commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,107 @@
+#endif
+
+// defines to avoid typecasts from (char_u *) to (char *) and back
+// (vim_strchr() and vim_strrchr() are now in alloc.c)
```
the alloc.c comment is outdated.

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,107 @@
+#define vim_strpbrk(s, cs) (char_u *)strpbrk((char *)(s), (char *)(cs))
+
+// FIXME(SplinterOfChaos): the strdup functions should be used without the x-
+// prefix, but even when the feature test passes, they may cause errors.
```
Has this been dealt with? I'm not sure what "errors" is referring to.

splinterofchaos replied:
At first, I wanted to just make these functions `strdup()`, `stpcpy()`, etc., to make our code look more like standard C. I can't remember exactly what error occurred, but I thought it'd be less confusing to people on platforms that don't have `strdup()` if we kept the `x` prefix, anyway. I removed this comment.

Pyrohh commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,107 @@
+#elif defined(HAVE_STRNICMP)
+# define STRNICMP(d, s, n) strnicmp((char *)(d), (char *)(s), (size_t)(n))
+#elif defined(HAVE_U_STRNICMP)
+# define STRICMP(d, s)     _strnicmp((char *)(d), (char *)(s))
```
Are you sure this has the correct amount of arguments? I'm confused after looking here:
https://msdn.microsoft.com/en-us/library/chd90w8e.aspx

splinterofchaos replied:
I just noticed that, too. Fixed.

splinterofchaos commented:
Sure thing. Rebased again.

Pyrohh commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,104 @@
+#elif defined(HAVE_STRICMP)
+# define STRICMP(d, s)     stricmp((char *)(d), (char *)(s))
+#elif defined(HAVE_U_STRICMP)
+# define STRICMP(d, s)     _stricmp((char *)(d), (char *)(s))
```
This and the above #define are misaligned with the first.

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,104 @@
+# define xstrlcpy(a, b, s) strlcpy(a, b, s)
+#endif
+#ifdef HAVE_MEMRCHR
+# define xmemrchr(s, c) memrchr(s, c)
```
Could you align this block? I ask because this was done for the above blocks, e.g. the one starting with `#define STRLEN`.

splinterofchaos replied:
:+1:

Pyrohh commented:
Besides the (very minor) nits, this looks great :-)

justinmk commented on src/nvim/os/env.c:
```diff
@@ -27,12 +27,23 @@
+  return strempty(e) ? NULL : e;
+}
+
+/// Returns whether an environment variable has been set, even if empty.
```
> Returns `true` if the environment variable is defined (even if empty).

splinterofchaos replied:
:+1:

justinmk commented on src/nvim/os/env.c:
```diff
@@ -557,9 +553,6 @@ void home_replace(buf_T *buf, char_u *src, char_u *dst, int dstlen, bool one)
   homedir_env_orig = homedir_env = (char_u *)os_getenv("HOME");
-  /* Empty is the same as not set. */
-  if (homedir_env != NULL && *homedir_env == NUL)
-    homedir_env = NULL;
```
Logic isn't entirely the same. Before, `homedir_env_orig` is later used to decided whether to free `homedir_env`. I haven't completely understood the logic, but it might be worth a closer look, or maybe `homedir_env_orig` can be eliminated.

oni-link replied:
A call to `modify_fname()` could change `homedir_env`. If the pointer is changed, it needs to be freed.

The result for `homedir_env_orig = homedir_env = ""` and `homedir_env_orig = homedir_env = NULL` should be the same. The only difference in the codepath is setting `envlen` to `0` again.

splinterofchaos replied:
I replaced `homedir_env_orig` with `bool must_free`. It looks to me like `envlen` will have the same value as the old code when `$HOME = ""`.

justinmk commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,104 @@
+#endif
+
+/// Compares two strings for equality.
+static inline bool strequal(const char *a, const char *b)
```
Since we have the `ascii_` family of functions now, wouldn it make sense to name this `asciiequal`, `asciinequal`, etc.?

justinmk commented on src/nvim/memory.h:
```diff
@@ -2,6 +2,104 @@
 
 #include <stddef.h>  // for size_t
+#include <string.h>
+#include <strings.h>
```
What's this for? I do not think this is available in MSVC, plus some comments suggest it is non-standard.

equalsraf replied:
There is a guard HAVE_STRINGS_H

splinterofchaos replied:
Right, it's for `strcasecmp` only. I could wrap it with `HAVE_STRINGS_H`, but I chose `HAVE_STRCASECMP` instead since it seems more topical.

justinmk commented on src/nvim/memory.c:
```diff
@@ -286,10 +286,11 @@ size_t memcnt(const void *data, char c, size_t len)
-/// The xstpcpy() function shall return a pointer to the terminating NUL
+/// The stpcpy() function shall return a pointer to the terminating NUL
 /// character copied into the dst buffer. This is the only difference with
 /// strcpy(), which returns dst.
```
xstpcpy is different than stpcpy.

splinterofchaos replied:
:+1: -- and the other comments I removed the `x` prefix from.

aktau replied:
> xstpcpy is different than stpcpy.

I do remember that despite misgivings about the exact implementation when we did it earlier, we did try to follow the standard. How is it different?

justinmk replied:
@aktau I misread, see below.

aktau replied:
:+1: :)

oni-link commented on config/CMakeLists.txt:
```diff
@@ -21,6 +21,7 @@ check_include_files(langinfo.h HAVE_LANGINFO_H)
 check_include_files(libgen.h HAVE_LIBGEN_H)
 check_include_files(locale.h HAVE_LOCALE_H)
 check_include_files(pwd.h HAVE_PWD_H)
+check_include_files(string.h HAVE_STRING_H)
```
Why do we have to check for `string.h`? This should be a C99 header.

splinterofchaos replied:
I added it with the check for `strings.h`, but yeah, it's redundant, so I deleted it.

oni-link commented on src/nvim/eval.c:
```diff
@@ -15081,16 +15080,15 @@ static int get_winnr(tabpage_T *tp, typval_T *argvar)
     if (arg == NULL)
       nr = 0;                   /* type error; errmsg already given */
-    else if (STRCMP(arg, "$") == 0)
+    else if (strequal(arg, "$") == 0)
```
`== 0` needs to be removed.

splinterofchaos replied:
:+1:

oni-link commented on src/nvim/eval.c:
```diff
@@ -15081,16 +15080,15 @@ static int get_winnr(tabpage_T *tp, typval_T *argvar)
+    else if (strequal(arg, "$") == 0)
       twin = (tp == curtab) ? lastwin : tp->tp_lastwin;
-    else if (STRCMP(arg, "#") == 0) {
+    else if (strequal(arg, "#") == 0) {
```
`== 0` needs to be removed.

splinterofchaos replied:
:+1:

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -1659,7 +1659,7 @@ static char_u *viminfo_filename(char_u *file)
 #ifdef VIMINFO_FILE2
       /* don't use $HOME when not defined (turned into "c:/"!). */
-      if (os_getenv((char_u *)"HOME") == NULL) {
+      if (os_getenv("HOME") == NULL) {
```
`os_getenv("HOME") == ""` could now set file to `VIMINFO_FILE2`.

splinterofchaos replied:
:+1:  -- Changed it to use `os_env_exists()`.

splinterofchaos commented:
I think I've address the concerns, but I'll wait for another LGTM or two before merging. Thanks all!

justinmk commented:
My comment at https://github.com/neovim/neovim/pull/2574#discussion_r30866448 is to raise the question about whether it is correct to replace all xstpcpy with stpcpy. `xstpcpy` behaves differently. I am not sure if the different behavior is used anywhere yet, but it doesn't seem right to do a 1-to-1 replacement. Perhaps that should be done in a separate PR.

splinterofchaos commented:
@justinmk I'm not aware of any functional differences, except that we marke `dst` and `src` with `restrict`. Though, if you look at [implementation at  opensource.apple.com](http://www.opensource.apple.com/source/Libc/Libc-320/string/FreeBSD/stpcpy.c), if `dst` and `src` did overlap and `dst > src`, the algorithm would never terminate since the `NUL` would be overwritten.

justinmk commented:
Oh man, I misread that comment. 

```
 /// The xstpcpy() function shall return a pointer to the terminating NUL
 /// character copied into the dst buffer. This is the only difference with
 /// strcpy(), which returns dst.
```

I thought it was describing a difference between `xstpcpy` and `cstpcpy`, *not* `strcpy`. That's really confusing, you were right to change the comment (xstpcpy => stpcpy).

Other than that LGTM

splinterofchaos commented:
Rebased. I plan to merge after the tests pass.

>  That's really confusing, you were right to change the comment (xstpcpy => stpcpy).

Well, later down in that comment block:
```c
/// This is the Neovim version of stpcpy(3) as defined in POSIX 2008. We
/// don't require that supported platforms implement POSIX 2008, so we
/// implement our own version.
```
I think it makes sense to refer to it as `xstpcpy` in the documentation since it differentiates it from the POSIX function. I originally wanted to remove the `x` prefixes from these functions, but it caused problems (don't remember exactly) and also felt wrong as they might be defined as macros for `__builtin_stuff`, but I forgot to revert the documentation.

justinmk commented on src/nvim/message.c:
```diff
@@ -743,7 +743,7 @@ void ex_messages(exarg_T *eap)
 
   s = os_getenv("LANG");
-  if (s != NULL && *s != NUL)
+  if (!s)
```
This should be `if (s)`

splinterofchaos replied:
Uck, I've made several one-off errors in this PR, but the unit tests caught none of them. Fixed, but on closer inspection, why do we only print "Message Maintainer: ..." if `$LANG` happens to be set, and not use it?

justinmk replied:
The "Message Maintainer" message will be removed soon anyways (what the hell is a messages maintainer?). I didn't realize this was for that :)

splinterofchaos commented:
Legacy tests on mac are failing on travis after a couple retries, but they pass locally. Could be a travis issue, so I'd appreciate if someone on a mac could try them, but for now, I'll assume that I need to go once more over this and make sure everything's right.

Pyrohh commented:
@splinterofchaos is there anything in particular blocking this PR? Besides what you mentioned above of course (silly Travis).

splinterofchaos commented:
@Pyrohh I'm just not 100% that I haven't caused the travis failures. The mac builds have been stalling while running old-style tests, and while they often fail on other PRs, it seems like different issues--the screen being out-of-sync, profiling information somehow getting mixed in with `screen:expect()`, and things like that. 

Right now, I'm just trying to narrow down what's safe. If the `os_getenv()` commit works (restarted the tests), I'll probably merge it manually, then try and introduce the other changes more granually.

splinterofchaos commented:
Yup, the `os_getenv()` commit passed and I manually merged it, so now I just have to figure out if anything messed up on the `strings` commits. i'll probably be totally reworking the changes.

Pyrohh commented:
Thanks @splinterofchaos.

justinmk commented:
related: https://github.com/neovim/neovim/pull/6460

blueyed commented:
> Yup, the `os_getenv()` commit passed and I manually merged it, so now I just have to figure out if anything messed up on the `strings` commits. i'll probably be totally reworking the changes.

I think that is rather bad, and should be reverted maybe.
While you can use `os_env_exists()` that is just extra overhead.
Ref: https://github.com/neovim/neovim/pull/10655#pullrequestreview-268248658

justinmk commented:
`str[i]equal` were merged separately. `strempty` would be worth including. The changes to `src/nvim/eval.c` are not needed, so we can close this.

> I think that is rather bad, and should be reverted maybe.

What is bad exactly?

> While you can use `os_env_exists()` that is just extra overhead.

Overhead compared to what?

blueyed commented:
@justinmk 
I've meant that is is rather confusing that NULL means empty value, and that you have to check for if it exists then additionally.
This is kind of reflecting the behavior Vim/Windows has..

justinmk commented:
True. I don't care much either way, but the change was meant as a convenience since most cases always check `e != NULL && *e != NUL`.
